### PR TITLE
Remove duplicate implementation of Collatz sequence

### LINKED ISSUE #8061: Remove duplicate Collatz sequence implementation
### Feature description

I think one of [`3n_plus_1.py`](https://github.com/TheAlgorithms/Python/blob/master/maths/3n_plus_1.py) and [`collatz_sequence.py`](https://github.com/TheAlgorithms/Python/blob/master/maths/collatz_sequence.py) should be removed, as they're nearly identical implementations of the same algorithm for computing Collatz sequences.

I'm not sure which of the two should be deleted, as I'm not sure which of "the Collatz conjecture" and "the 3n + 1 problem" is the more well-known name for the problem.

### PR DESCRIPTION
**Pull request by tianyizheng02**
### Describe your change:

Fixes #8061

- Removed `maths/3n_plus_1.py` in favor of `maths/collatz_sequence.py` because the latter is the more well-known name for the problem
- Added tests from `maths/3n_plus_1.py` to `maths/collatz_sequence.py`
- Converted the `collatz_sequence` function to save memory
- Expanded explanation of the Collatz conjecture

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [x] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [ ] All filenames are in all lowercase characters with no spaces or dashes.
* [ ] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

File: maths/collatz_sequence.py [hunk 1]
```diff
@@ -1,43 +1,64 @@
...
-    sequence = [n]
+    yield n
     while n != 1:
-        n = 3 * n + 1 if n & 1 else n // 2
-        sequence.append(n)
-    return sequence
+        if n % 2 == 0:
+            n //= 2
+        else:
+            n = 3 * n + 1
```
Comment (cclauss) [inline maths/collatz_sequence.py:53]: # OPTIONAL:
```suggestion
        div, mod = divmod(n, 2)
        n = div if mod else 3 * n + 1
```

**Review by cclauss — CHANGES_REQUESTED** [review]
Looking cool!  Do these changes effect performance?

**Comment by tianyizheng02** [comment]
> Looking cool! Do these changes effect performance?

@cclauss I don't think performance is really affected in most cases since the algorithm itself is basically unchanged. However, I think there'd be slightly better performance if one were to iterate over the algorithm's output:
```py
for num in collatz_sequence(n): 
    # do something with num
```
since having a generator means not needing to create and store the entire sequence in memory. However, I haven't timed this generator implementation against the original list implementation, so I don't know how much of a difference it makes.

**Review by cclauss — APPROVED** [review]
Nice!

**Pull request merged**